[Diatbbin] ip

Diatbbin opened the pull request:
# DukePro
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

DukePro frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ _SUPER_ FAST to use

All you need to do is,

1.  download it from [here](https://github.com/nus-cs2103-AY2223S2/ip/pull/137).
2.  double-click it.
3.  add your tasks.
4.  let it manage your tasks for you 😉

And it is **FREE**!

Features:

- [x] Managing tasks
- [ ]  Managing deadlines (coming soon)
- [ ]  Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the `main` method:

```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Diatbbin committed:
Completed Level-1

Diatbbin committed:
Completed Level-2

Diatbbin committed:
Completed Level-3

Diatbbin committed:
Completed Level-4: ToDos, Events, Deadlines

Diatbbin committed:
Added A-TextUiTesting: Automated Text UI Testing

Diatbbin committed:
Completed Level-5: Handle Errors

Diatbbin committed:
Completed Level-6: Delete

Diatbbin committed:
Added exception handling for mark tasks

Diatbbin committed:
Change exception handling code structure

Diatbbin committed:
Remove static field count from Task

Diatbbin committed:
Replace if else with switch case

Diatbbin committed:
Complete level 7 save

Diatbbin committed:
Complete level 8 dates and times

Diatbbin committed:
Merge branch 'branch-Level-7'

Diatbbin committed:
Merge branch 'branch-Level-8'

Diatbbin committed:
Refactor code to use Ui and Parser class

Diatbbin committed:
Refactor code to use TaskList class

Diatbbin committed:
Complete A-MoreOOP and clean code

Diatbbin committed:
Organize files into Packages

Diatbbin committed:
Merge remote-tracking branch 'origin/add-gradle-support'

Diatbbin committed:
Set up gradle for Duke

Diatbbin committed:
Add JUnit tests

Diatbbin committed:
Debug FileSystem class and clean up code

Diatbbin committed:
Package the app as a JAR file

Diatbbin committed:
Correct Duke's name

Diatbbin committed:
Add JavaDoc comments

Diatbbin committed:
Tweak code to follow the coding standard

Diatbbin committed:
Add search functionality

Diatbbin committed:
Merge branch 'branch-A-JavaDoc'

Diatbbin committed:
Merge branch 'branch-A-CodingStandard'

Diatbbin committed:
Merge branch 'branch-Level-9'

Diatbbin committed:
Refactor Task subclasses

Diatbbin committed:
Implement checkstyle

Diatbbin committed:
Fix checkstyle violations

Diatbbin committed:
Merge branch 'branch-A-CheckStyle'

Diatbbin committed:
Add GUI for duke

Diatbbin committed:
Merge branch 'branch-Level-10'

Diatbbin committed:
Fix checkstyle violations

adam07018 commented on src/main/java/duke/helper/TaskList.java:
```diff
@@ -0,0 +1,78 @@
+     * @return task that has been deleted
+     */
+    public Task deleteTask(int taskNo) {
+        Task task = tasks.get(taskNo);
```
Is it better to add a try catch block because there might be possible indexoutofbound error

adam07018 commented on src/main/java/duke/helper/TaskList.java:
```diff
@@ -0,0 +1,78 @@
+     * @param taskId id of the task
+     */
+    public String mark(boolean isDone, String taskId) {
+        int taskNo = Integer.parseInt(taskId) - 1;
```
Maybe can add try catch because user might input something else except number:cowboy_hat_face:

adam07018 commented on src/main/java/duke/helper/Parser.java:
```diff
@@ -0,0 +1,149 @@
+            return Ui.filter(output);
+
+        case "bye":
+            fileSystem.updateFile(tasks);
```
:thought_balloon:Is it necessary to update when bye command is triggered?

adam07018 commented on src/main/java/duke/helper/Parser.java:
```diff
@@ -0,0 +1,149 @@
+            int month = Integer.parseInt(period[1]);
+            int day = Integer.parseInt(period[0]);
+            int hour = Integer.parseInt(yearTime[1].substring(0, 2));
+            int min = Integer.parseInt(yearTime[1].substring(2));
```
Suggest to use a formatter which make life much easier:innocent:

adam07018 commented on src/main/java/duke/helper/Ui.java:
```diff
@@ -0,0 +1,98 @@
+ * Ui class that handles all the interactions with the users
+ */
+public class Ui {
+    private final String line = "_______________________________________________________________________";
```
Sorry I don't get the use of this:thinking:

adam07018 commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,46 @@
+     */
+    public Deadline(String description, String by) throws InvalidDateTimeException {
+        super(description, false, "D");
+        this.dueDateTime = Parser.handleDateTime(by).toString();
```
Can we do this in Parser class?

adam07018 commented on src/main/java/duke/helper/Parser.java:
```diff
@@ -0,0 +1,149 @@
+     * @return a LocalDateTime object
+     * @throws InvalidDateTimeException If incorrect dateTime values are given
+     */
+    public static LocalDateTime handleDateTime(String dateTime) throws InvalidDateTimeException {
```
Should we create a time object class?

adam07018 commented on src/main/java/duke/storage/FileSystem.java:
```diff
@@ -0,0 +1,117 @@
+     * @throws FileNotFoundException If the file cannot be found
+     */
+    public ArrayList<Task> loadFromFile() throws FileNotFoundException, InvalidDateTimeException {
+        Scanner scanner = new Scanner(file);
```
Nice function!

adam07018 reviewed (commented):
Review comment by adam07018

Diatbbin committed:
Add dateTime formatter and clean up code

SeeuSim commented on src/main/java/duke/helper/Parser.java:
```diff
@@ -0,0 +1,144 @@
+     * @param desc description of the task
+     * @return String representation of Task
+     * @throws DukeException If input is invalid
+     */
```
It seems that a lot of logic is crammed into this method.

In the event that this method bugs out due to one of the methods, it can be difficult to find the error point and the offending switch case. 

I would suggest encapsulating the logic for each switch block in its own `Command` class, and all of them may have a `simulate` method that handles their own errors and outputs a UI regardless. This also makes it easy to define test cases for each of the commands and isolate the logic there.
Also, this would in the future make it easier to change the logic for each command, as you would only need to modify the target Command class' `simulate` method.

SeeuSim commented on src/main/java/duke/helper/Ui.java:
```diff
@@ -0,0 +1,86 @@
+    }
+
+    public static String showErrorMsg(String desc) {
+        return String.format("%s commands need to be followed by an integer!\n", desc.split(" ")[0]);
```
These can be abstracted into Exceptions, and all the Exceptions may be passed into a single UI method that is responsible for displaying Exceptions. You can then unit test that the exceptions are output correctly to the console.

SeeuSim commented on src/main/java/duke/helper/Ui.java:
```diff
@@ -0,0 +1,86 @@
+     * @param task task to be removed
+     * @param size number of remaining tasks
+     */
+    public static String showDelete(Task task, int size) {
```
These can be output by a Task -> DeleteTask class that is responsible for outputting the relevant UI message to the UI. The UI then need only have a single method that parses string messages and outputs them in the desired format, or even be changed to output UI elements in the GUI.

SeeuSim commented on src/main/java/duke/storage/FileSystem.java:
```diff
@@ -0,0 +1,117 @@
+     * @param taskList the TaskList that contains all the tasks that will be stored in the file
+     */
+    public void updateFile(TaskList taskList) {
+        try {
```
This can be expensive (rewriting the entire file with each operation) if your task list grows in size. Instead, you may modify specific rows/lines in your file with each read/write operation and keep your interactions with the file *atomic*

SeeuSim commented on src/main/java/duke/storage/FileSystem.java:
```diff
@@ -0,0 +1,117 @@
+        ArrayList<Task> tasks = new ArrayList<>();
+
+        while (scanner.hasNextLine()) {
+            String task = scanner.nextLine();
```
this logic that is repeated in a loop may be abstracted into its own function/class so that in the event of an error, it is easy to find and fix.

SeeuSim reviewed (commented):
In general, good approaches are used in the project, but they may be further improved by:
- code splitting: Breaking the logic up into atomic chunks that handle their own errors and are easy to modify and fix.
- more test cases: Currently, events such as invalid dates being passed to Events or wrong usage of creating Events/Deadlines are not tested, and if future changes break the code, the logic will fail silently.

Diatbbin committed:
Add assertions

There are several assumptions made about the program state that
could break the program

Use assertions to define these assumptions so that the runtime
can verify them

Let's
* add assertions to ensure these assumptions are not violated

Diatbbin committed:
Improve code quality

There are long methods that affect readability

Keeping a method less than 30 lines makes it easier to find a bug

Let's,
* Use functions to reduce the number of lines used for a method

Diatbbin committed:
Merge pull request #2 from Diatbbin/branch-A-Assertions

Add assertions

Diatbbin committed:
Merge branch 'master' into branch-A-CodeQuality

Diatbbin committed:
Merge pull request #3 from Diatbbin/branch-A-CodeQuality

Improve code quality

Diatbbin committed:
Add more exceptions

Diatbbin committed:
Add C-help

Diatbbin committed:
Merge branch 'branch-BCD-Extension'

Diatbbin committed:
Fix error msgs not showing on GUI bug and edit dateTime formatter

Diatbbin committed:
Update unit tests

Diatbbin committed:
Add user guide

Diatbbin committed:
Rename screenshot of product to Ui.png

Diatbbin committed:
Update user guide

Diatbbin committed:
Fix checkstyle violations

Diatbbin committed:
Fix bugs with commands and clean up code

Diatbbin committed:
Improve the GUI

Diatbbin committed:
Update Ui.png

Diatbbin committed:
Update user guide

Diatbbin committed:
Update javadocs and clean up code

Diatbbin committed:
Fix GUI bug where list cant display all the stored tasks

Diatbbin committed:
Update Ui.png

Diatbbin committed:
Update user guide

Diatbbin committed:
Remove redundant statements

Diatbbin committed:
Fix checkstyle violations

Diatbbin committed:
Clean up code and improve naming

Diatbbin committed:
Fix bugs related to missing inputs

Diatbbin committed:
Update user guide

Diatbbin committed:
Add credits

Diatbbin committed:
Update assertions

Diatbbin committed:
Update assertions for task number inputs

Diatbbin committed:
Update variable names in TaskList
